### PR TITLE
Test update with row scaling

### LINKED ISSUE #2585: Create test for `analysis::run_analysis_update_with_rowscaling` in: update.cpp
Related to: #2574

### PR DESCRIPTION
**Pull request by dafeda**
**Issue**
Resolves #2585 

**Approach**

Seems like the distance between posterior and ML-estimate when obs_std is very low does not approach 0 with row scaling, i.e., this test fails:

In this specific test I set alpha for the first row to 1.0 and to 0.9 for the second row.

```cpp
    // THEN("At strong beliefs, we should be close to the ml-estimate") {
    //     REQUIRE(d_posterior_ml[n_sd - 1] < eps);
    // }
```

## Pre review checklist

- [x ] Added appropriate labels

Adding labels helps the maintainers when writing release notes, see sections and the
corresponding labels here: https://github.com/equinor/ert/blob/main/.github/release.yml

### CODE REVIEW

File: libres/tests/analysis/test_update.cpp [hunk 1]
```diff
@@ -46,16 +53,16 @@ struct model {
...
-SCENARIO("Running analysis update without row scaling on linear model",
+SCENARIO("Running analysis update with and without row scaling on linear model",
          "[analysis]") {
 
     GIVEN("Fixed prior and measurements") {
 
         analysis_module_type *enkf_module = analysis_module_alloc("STD_ENKF");
         analysis_module_set_var(enkf_module, ENKF_TRUNCATION_KEY_, "1.0");
-        rng_type *rng = rng_alloc(MZRAN, INIT_DEFAULT);
+        rng_type *rng = rng_alloc(MZRAN, INIT_DEV_URANDOM);
```
Comment (joakim-hove) [inline libres/tests/analysis/test_update.cpp:63]: I would stick to `INIT_DEFAULT` in order to get reproducible results

Comment (dafeda) [reply libres/tests/analysis/test_update.cpp:63]: The idea is to write the type of tests that are agnostic to the seed, i.e., test properties that should hold no matter the seed.

File: libres/tests/analysis/test_update.cpp [hunk 2]
```diff
@@ -200,6 +207,193 @@ SCENARIO("Running analysis update without row scaling on linear model",
...
+                obs_block_iset(ob, iobs, measurements[iobs], 1.0);
+            }
+            matrix_type *E = obs_data_allocE(obs_data, rng, ens_size);
+
+            analysis::run_analysis_update_with_rowscaling(
+                enkf_module, ens_mask, meas_data, obs_data, rng, E, parameters);
+
+            THEN("Updated parameter matrix should equal prior parameter "
+                 "matrix") {
+                REQUIRE(matrix_similar(A, A_with_scaling, 1e-12));
```
Comment (joakim-hove) [inline libres/tests/analysis/test_update.cpp:232]: This is actually one of few situations where you could use `matrix_equal()` for bitwise comparison.

File: libres/tests/analysis/test_update.cpp [hunk 3]
```diff
@@ -200,6 +207,193 @@ SCENARIO("Running analysis update without row scaling on linear model",
                 REQUIRE(d_posterior_ml[n_sd - 1] < eps);
             }
         }
+
+        WHEN("Row scaling factor is 0 for both parameters") {
+            row_scaling_type *row_scaling = row_scaling_alloc();
```
Comment (joakim-hove) [inline libres/tests/analysis/test_update.cpp:212]: For these `xxx_type * ptr = xxx_alloc()` calls you can choose to use `auto * ptr = xxx_alloc(); - I guess that would be more idiomatic C++.

File: libres/tests/analysis/test_update.cpp [hunk 4]
```diff
@@ -200,6 +207,193 @@ SCENARIO("Running analysis update without row scaling on linear model",
...
+            analysis::run_analysis_update_with_rowscaling(
+                enkf_module, ens_mask, meas_data, obs_data, rng, E, parameters);
+            analysis::run_analysis_update_without_rowscaling(
+                enkf_module, ens_mask, meas_data, obs_data, rng, E,
+                A_no_scaling);
+
+            THEN("First row of scaled parameters should equal first row of "
+                 "unscaled parameters, while second row of scaled parameters "
+                 "should equal prior") {
+                matrix_type *A_with_scaling_T =
```
Comment (joakim-hove) [inline libres/tests/analysis/test_update.cpp:298]: Here you can use:
```
auto * A_with_scaling_T = matrix_alloc_transpose(A_with_scaling);
```

File: libres/tests/analysis/test_update.cpp (hunk 1 above)
Comment (joakim-hove) [reply libres/tests/analysis/test_update.cpp:63]: I see the point - but you can not actually verify that you have succeeded in that; other than that flakyness would imply that the test is not agnostic to seed.

**Comment by joakim-hove** [comment]
> Seems like the distance between posterior and ML-estimate when obs_std is very low does not approach 0 with row scaling, i.e., this test fails:

I am not fully up to speed on what fails here, but mind that a row scaling value < 1 can loosely speaking be seen as an inflated observation std, i.e. the "effective std" is not "very low".

File: libres/tests/analysis/test_update.cpp (hunk 2 above)
Comment (dafeda) [reply libres/tests/analysis/test_update.cpp:232]: > > Seems like the distance between posterior and ML-estimate when obs_std is very low does not approach 0 with row scaling, i.e., this test fails:
> 
> I am not fully up to speed on what fails here, but mind that a row scaling value < 1 can loosely speaking be seen as an inflated observation std, i.e. the "effective std" is not "very low".

Thanks, I think that makes sense.

Comment (dafeda) [reply libres/tests/analysis/test_update.cpp:232]: > > Seems like the distance between posterior and ML-estimate when obs_std is very low does not approach 0 with row scaling, i.e., this test fails:
> 
> I am not fully up to speed on what fails here, but mind that a row scaling value < 1 can loosely speaking be seen as an inflated observation std, i.e. the "effective std" is not "very low".

Thanks, I think that makes sense.

**Review by joakim-hove — APPROVED** [review]

File: libres/tests/analysis/test_update.cpp (hunk 3 above)
Comment (dafeda) [reply libres/tests/analysis/test_update.cpp]: Thanks, I agree.

File: libres/tests/analysis/test_update.cpp (hunk 4 above)
Comment (dafeda) [reply libres/tests/analysis/test_update.cpp]: Ah, that's better, thanks.

File: libres/tests/analysis/test_update.cpp (hunk 1 above)
Comment (dafeda) [reply libres/tests/analysis/test_update.cpp]: After some thought, I agree with you and have switched back to INIT_DEFAULT.
Thanks!

**Comment by dafeda** [comment]
jenkins: test this please

**Pull request merged**